From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Triassic <techbyteofficial9@gmail.com>
Date: Sun, 12 May 2024 17:33:01 +0300
Subject: [PATCH] Empty configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index bbc4d7d3ca84642828f9a3f788ca26bba900d15b..679124845dea42f0910ea2cab2638a13de0780ec 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2251,6 +2251,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // LinearPaper start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLinearConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // LinearPaper end
+
         /**
          * Sends the component to the player
          *
