From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Encode42 <me@encode42.dev>
Date: Thu, 16 Sep 2021 20:32:04 -0400
Subject: [PATCH] Tentacles config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 88767865cc2dce0e341eef0f53323af6f95fb841..1a9a771fc75d9a98a794a57c6cfc752d89f149f2 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1694,6 +1694,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Tentacles start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getTentaclesConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Tentacles end
+
         /**
          * Sends the component to the player
          *
