From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Triassic <techbyteofficial9@gmail.com>
Date: Sun, 12 May 2024 17:33:01 +0300
Subject: [PATCH] Empty configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index c043d239c449bf4bb13a24467f2f6c67b4d28d2d..cb812f8cc17af4c2ddf660afc440a79487728179 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2254,6 +2254,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // LinearPaper start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLinearConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // LinearPaper end
+
         /**
          * Sends the component to the player
          *
