From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Triassic <techbyteofficial9@gmail.com>
Date: Sun, 12 May 2024 17:33:01 +0300
Subject: [PATCH] Empty configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 5aa64ea39ebd92e5067c53cea49a8685c0b9eee4..6667f35c6cd09d1c833aed90e42d8a20430bc0ee 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2265,6 +2265,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // LinearPaper start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLinearConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // LinearPaper end
+
         /**
          * Sends the component to the player
          *
