From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Encode42 <me@encode42.dev>
Date: Wed, 14 Jul 2021 19:22:33 -0400
Subject: [PATCH] Tentacles config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index b8c380dae58d47514d684a6de2fd17b7d5135ab1..02dac001cce2c0bf4ff6a29f49d0c957db2a6d5a 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1700,6 +1700,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Tentacles start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getTentaclesConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Tentacles end
+
         /**
          * Sends the component to the player
          *
