From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Encode42 <me@encode42.dev>
Date: Thu, 16 Sep 2021 20:32:04 -0400
Subject: [PATCH] Tentacles config files


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ccfb8a8dcd51bf998c79298a22e38cebb2be6d39..b12c575310d19547cf8f9c3d9ca043ee720cd9c9 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1730,6 +1730,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Tentacles start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getTentaclesConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Tentacles end
+
         /**
          * Sends the component to the player
          *
